### ISSUE TITLE
pre signed upload url not working when adding Metadata

### ISSUE THREAD

**Issue by aymanov1993** [minio/minio#5857]
Hi

I tried to upload a file to minio server using php s3client pre signed url with adding Metadata by "Metadata" s3client option and I got the following response :  

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
    <Code>SignatureDoesNotMatch</Code>
    <Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>
    <Key></Key>
    <BucketName></BucketName>
    <Resource>/tenant0/public/cv/other/07/1524560347.jpg</Resource>
    <RequestId>3L137</RequestId>
    <HostId>3L137</HostId>
</Error>
```

My code :

```php
$cmd = $this->s3Client->getCommand('PutObject', [
            'Bucket' => $bucket,
            'Key'    => $fileName,
            'Metadata' => [
                "A" => "B",
            ],
        ]);
       
        $request = $this->s3Client->createPresignedRequest($cmd, "+${minutes} minutes");
        
        // Get the actual presigned-url
        $presignedUrl = (string) $request->getUri();
        return $presignedUrl;
```

Request url :

http://192.168.6.5:9000/tenant0/public/cv/other/07/1524560347.jpg?x-amz-meta-A=B&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=Y7NPQFOYR18YH8Y8XKSJ%2F20180424%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20180424T085907Z&X-Amz-SignedHeaders=host&X-Amz-Expires=1200&X-Amz-Signature=c5121c4e287d4cbd6840165f4a718852ec403d6f9ded0c8ec82b4131e036d5ca

What should I do to adding a custom data to be returned with the upload event handler ?

## Expected Behavior

Adding Metadata when generate a pre signed upload URL in php using "aws/aws-sdk-php 3.54" SDK shouldn't corrupt the signature.

## Current Behavior

Corrupted signature when generate a pre signed upload URL in php using "aws/aws-sdk-php 3.54" SDK shouldn't corrupt the signature.

## Steps to Reproduce (for bugs)

Described in the summary.

## Context

Adding extra data to get it on uploading event notification like webhook

## Your Environment

* Version used (`minio version`): 23
* Environment name and version : php with AWS SDK
* Server type and version: centos 6.7

**Comment by harshavardhana** [comment]
This is a client bug @aymanov1993 aws-sdk-php is not canonicalizing the query params properly, instead, it is signing `x-am-meta` in its lowercase form whereas the HTTP RFC indicates that it should be in canonical form.

We could fix it in our code by doing both forms just to make it work, but it is not correct and should be fixed on the client as well.

**Comment by michaelhyolee** [comment]
+1 this is stopping us from using Minio as a drop-in replacement for S3 (using aws-sdk-ruby V2)

**Comment by atrauzzi** [comment]
Likewise, very blocked by this now.

**Comment by atrauzzi** [comment]
Is there any timeline on when this fix will be available?

I just tried the latest docker image and I'm still unable to upload, getting the message ` There were headers present in the request which were not signed` from the AWS PHP SDK.

Only seems to happen when I try to include metadata with my files.

**Comment by harshavardhana** [comment]
This should be already fixed @atrauzzi are you sure you are setting the x-amz headers correctly in your client request?

**Comment by kannappanr** [comment]
@atrauzzi Here is a test I did with `mc` to prove that `minio` already handles this. I created a presigned url using `mc`
```
$ mc share upload play/tmptest/file.txt
URL: https://play.minio.io:9000/tmptest/file.txt
Expire: 7 days 0 hours 0 minutes 0 seconds
Share: curl https://play.minio.io:9000/tmptest/ -F x-amz-signature=6dc7019cf0f8953ff35bb701f0f342d97982d31588f8cbac4f7cbd2705a55f5c -F bucket=tmptest -F policy=eyJleHBpcmF0aW9uIjoiMjAxOC0wOC0wNVQwODozMjo1MS44NjlaIiwiY29uZGl0aW9ucyI6W1siZXEiLCIkYnVja2V0IiwidG1wdGVzdCJdLFsiZXEiLCIka2V5IiwiZmlsZS50eHQiXSxbImVxIiwiJHgtYW16LWRhdGUiLCIyMDE4MDcyOVQwODMyNTJaIl0sWyJlcSIsIiR4LWFtei1hbGdvcml0aG0iLCJBV1M0LUhNQUMtU0hBMjU2Il0sWyJlcSIsIiR4LWFtei1jcmVkZW50aWFsIiwiUTNBTTNVUTg2N1NQUVFBNDNQMkYvMjAxODA3MjkvdXMtZWFzdC0xL3MzL2F3czRfcmVxdWVzdCJdXX0= -F x-amz-algorithm=AWS4-HMAC-SHA256 -F x-amz-credential=Q3AM3UQ867SPQQA43P2F/20180729/us-east-1/s3/aws4_request -F x-amz-date=20180729T083252Z -F key=file.txt -F file=@<FILE>
```
I added a `custom meta` header called `x-amz-meta-testtag` while doing an upload using th generated presigned url.
```
curl https://play.minio.io:9000/tmptest/ -F x-amz-signature=6dc7019cf0f8953ff35bb701f0f342d97982d31588f8cbac4f7cbd2705a55f5c -F bucket=tmptest -F policy=eyJleHBpcmF0aW9uIjoiMjAxOC0wOC0wNVQwODozMjo1MS44NjlaIiwiY29uZGl0aW9ucyI6W1siZXEiLCIkYnVja2V0IiwidG1wdGVzdCJdLFsiZXEiLCIka2V5IiwiZmlsZS50eHQiXSxbImVxIiwiJHgtYW16LWRhdGUiLCIyMDE4MDcyOVQwODMyNTJaIl0sWyJlcSIsIiR4LWFtei1hbGdvcml0aG0iLCJBV1M0LUhNQUMtU0hBMjU2Il0sWyJlcSIsIiR4LWFtei1jcmVkZW50aWFsIiwiUTNBTTNVUTg2N1NQUVFBNDNQMkYvMjAxODA3MjkvdXMtZWFzdC0xL3MzL2F3czRfcmVxdWVzdCJdXX0= -F x-amz-algorithm=AWS4-HMAC-SHA256 -F x-amz-credential=Q3AM3UQ867SPQQA43P2F/20180729/us-east-1/s3/aws4_request -F x-amz-meta-testtag=testtag -F x-amz-date=20180729T083252Z -F key=file.txt -F file=@/etc/issue
```

Here is the output of `mc stat` on that object.
```
kannappan@x1minio ~ $ mc stat play/tmptest/file.txt
Name      : file.txt
Date      : 2018-07-29 01:56:40 PDT 
Size      : 20B    
ETag      : 2fbcdb40c9edd1edc6c16ac2b0a401f8 
Type      : file 
Metadata  :
  Content-Type      : text/plain 
  X-Amz-Meta-Testtag: testtag 
```

**Comment by atrauzzi** [comment]
@harshavardhana - Ultimately, I'm using the AWS PHP SDK.  So I pass metadata to it and I assume it will handle things correctly, given that it's AWS who publishes the library. 😉 

At no point do I manually add headers, this is effectively what gets passed down to the AWS SDK in its options array:

```
        $options = [
            'Metadata' => array_merge(
                [
                    'organization_id' => $organizationId ?: '0',
                    'uploader_id' => $uploaderId,
                ],
                array_only($upload, [
                    'name',
                    'mime',
                ])
            ),
        ];
```

I think it's worth noting:
- I'm not explicitly creating a pre-signed URL.  I'm simply doing an upload.  So unless a pre-signed URL happens internally as part of this, I'm not sure why I'm getting this message from minio.
- If I comment out the entire metadata key, the requests work, with the obvious disadvantage at that point being that I can't set metadata on my objects.

---

Just to be sure, I pointed the exact same code at a real S3 bucket:

![image](https://user-images.githubusercontent.com/115199/43367875-9c2c263e-9319-11e8-8e15-9ac820534e5e.png)

No error, and my metadata is showing up as intended.

**Comment by atrauzzi** [comment]
@harshavardhana - Let's continue over at the new ticket I've opened, that way this doesn't get lost in a closed issue. 😟

**Comment by harshavardhana** [comment]
> @harshavardhana - Let's continue over at the new ticket I've opened, that way this doesn't get lost in a closed issue.

Thanks agreed @atrauzzi
